street_network/src/transform/sausage_links.rs

```diff
@@ -0,0 +1,162 @@
+use std::collections::BTreeSet;
+
+use crate::{
+    osm, BufferType, Direction, DrivingSide, LaneSpec, LaneType, OriginalRoad, StreetNetwork,
+};
+
+/// Find dual carriageways that split very briefly and then re-join, with no intermediate roads.
+/// Collapse them into one road with a barrier in the middle.
+pub fn collapse_sausage_links(streets: &mut StreetNetwork) {
+    for (id1, id2) in find_sausage_links(streets) {
+        fix(streets, id1, id2);
+    }
+}
+
+fn find_sausage_links(streets: &StreetNetwork) -> BTreeSet<(OriginalRoad, OriginalRoad)> {
+    let mut pairs: BTreeSet<(OriginalRoad, OriginalRoad)> = BTreeSet::new();
+
+    for (id1, road1) in &streets.roads {
+        // TODO People often forget to fix the lanes when splitting a dual carriageway, but don't
+        // attempt to detect/repair that yet.
+        if road1.oneway_for_driving().is_none() {
+            continue;
+        }
+        // Find roads that lead between the two endpoints
+        let mut common_roads: BTreeSet<OriginalRoad> =
+            into_set(streets.roads_per_intersection(id1.i1))
+                .intersection(&into_set(streets.roads_per_intersection(id1.i2)))
+                .cloned()
+                .collect();
+        // Normally it's just this one road
+        assert!(common_roads.remove(id1));
+        // If there's many roads between these intersections, something weird is happening; ignore
+        // it
+        if common_roads.len() != 1 {
+            continue;
+        }
+
+        let id2 = common_roads.into_iter().next().unwrap();
+        // Ignore if we've already found this match
+        if pairs.contains(&(id2, *id1)) {
+            continue;
+        }
+
+        let road2 = &streets.roads[&id2];
+        if road2.oneway_for_driving().is_none()
+            || road1.osm_tags.get(osm::NAME) != road2.osm_tags.get(osm::NAME)
+        {
+            continue;
+        }
+
+        // The two roads must point in a loop. Since they're both one-way, we can just
```

Comment:
Restoring @BudgieInWA's comments from https://github.com/a-b-street/abstreet/pull/933...

> Your "bad example" looks like a coherent collapsing of that topology, but in the wrong situation. It should be ok to represent a looping parking aisle as one way like that. The topology is the same as a sausage link, so we need to use more clues to classify it as one.
> The geometry is I think what we need to use. What if we linearly interpolate along both sides of the sausage, and take the distance between the two resulting points. We would expect all the values to be close for a dual carriage way, and get huge for loop roads.

Fortunately we didn't need to look at geometry to fix these cases. Just the graph structure was enough to tell. But the idea that both sides of a dual carriageway should stay roughly the same distance away from each other is a great one; I think it'll be necessary to handle cases like https://dabreegster.github.io/talks/asg_osm2streets/slides.html#/osm2streets-dual-carriageways-1.